tests/test_crawlers.py

```diff
@@ -599,3 +601,77 @@ def test_iterating(self):
+        of ftplib. Otherwise (encountering a filename) raise the proper exception """
+        if name not in ["..", "folder_name", ""]:
+            raise ftplib.error_perm
+
+    @mock.patch('geospaas_harvesting.crawlers.ftplib.FTP', autospec=True)
+    def test_ftp_correct_navigation(self, mock_ftp):
+        """ shall categorize the specific file names (based on specific 'fileformat' which is
+        revealed in the configuration file) as well as folder(s) inside the ftp resource """
+        test_crawler = crawlers.FTPCrawler('ftp:///', fileformat='.gz')
+        test_crawler.ftp.nlst.return_value = ['file1.gz', 'folder_name', 'file3.bb', 'file2.gz', ]
+        test_crawler.ftp.cwd = self.emulate_cwd_of_ftp
+        test_crawler.ftp.host = ''
+        with self.assertLogs('geospaas_harvesting.crawlers.FTPCrawler'):
+            test_crawler._explore_page('')
+        # '.gz' files must be in the "_urls" list
+        self.assertTrue(any([url_name.endswith('file1.gz') for url_name in test_crawler._urls]))
+        self.assertTrue(any([url_name.endswith('file2.gz') for url_name in test_crawler._urls]))
+        # Other type of files should not be in the "_urls" list
+        self.assertFalse(any([url_name.endswith('file3.bb') for url_name in test_crawler._urls]))
+        # folder with 'folder_name' must be in the "_to_process" list
+        self.assertTrue(any([url_name.endswith('folder_name')
+                             for url_name in test_crawler._to_process]))
+
+    # @mock.patch('geospaas_harvesting.crawlers.ftplib.FTP.socket')
+    @mock.patch('geospaas_harvesting.crawlers.ftplib.FTP.login')
+    @mock.patch('geospaas_harvesting.crawlers.ftplib.FTP.cwd')
+    def test_ftp_correct_exception(self, mock_cwd, mock_ftp):
+        """ shall return the costume 'ConnectionError'
+        (defined in raise_exception_reach_cwd_after_login) instead of 'ftplib.error_perm' in order
+        to continue the harvesting process in the case of redundant or repetitive login attempt(s)
+        after the first login attempt """
+        test_crawler = crawlers.FTPCrawler('ftp:///', username="d", password="d", fileformat='.gz')
+        mock_cwd.side_effect = self.raise_exception_reach_cwd_after_login
+        mock_ftp.side_effect = self.raise_exception_503
+        with self.assertRaises(ConnectionError):
+            with self.assertLogs('geospaas_harvesting.crawlers.FTPCrawler'):
+                test_crawler._explore_page('')
+        mock_ftp.side_effect = self.raise_exception_230
+        with self.assertRaises(ConnectionError):
+            with self.assertLogs('geospaas_harvesting.crawlers.FTPCrawler'):
+                test_crawler._explore_page('')
+        mock_ftp.side_effect = self.raise_exception_999
+        with self.assertRaises(ftplib.error_perm):
+            with self.assertLogs('geospaas_harvesting.crawlers.FTPCrawler'):
+                test_crawler._explore_page('')
```

Comment:
You can use exceptions directly as side effects, no need to define intermediate methods.

```suggestion
        mock_cwd.side_effect = ConnectionError
        mock_ftp.side_effect = ftplib.error_perm("503")
        with self.assertRaises(ConnectionError):
            with self.assertLogs('geospaas_harvesting.crawlers.FTPCrawler'):
                test_crawler._explore_page('')
        mock_ftp.side_effect = ftplib.error_perm("230")
        with self.assertRaises(ConnectionError):
            with self.assertLogs('geospaas_harvesting.crawlers.FTPCrawler'):
                test_crawler._explore_page('')
        mock_ftp.side_effect = ftplib.error_perm("999")
        with self.assertRaises(ftplib.error_perm):
            with self.assertLogs('geospaas_harvesting.crawlers.FTPCrawler'):
                test_crawler._explore_page('')
```

Reply:
Done in the next commit